relayer-cli/src/commands/tx/client.rs

```diff
@@ -67,50 +68,62 @@ pub struct TxUpdateClientCmd {
     trusted_height: Option<u64>,
 }
 
-impl Runnable for TxUpdateClientCmd {
-    fn run(&self) {
+impl TxUpdateClientCmd {
+    fn execute(&self) -> Result<IbcEvent, String> {
         let config = app_config();
 
-        let dst_chain = match spawn_chain_runtime(&config, &self.dst_chain_id) {
-            Ok(handle) => handle,
-            Err(e) => return Output::error(format!("{}", e)).exit(),
-        };
+        let dst_chain =
+            spawn_chain_runtime(&config, &self.dst_chain_id).map_err(|e| format!("{}", e))?;
 
-        let src_chain_id =
-            match dst_chain.query_client_state(&self.dst_client_id, ibc::Height::zero()) {
-                Ok(cs) => cs.chain_id(),
-                Err(e) => {
-                    return Output::error(format!(
-                        "Query of client '{}' on chain '{}' failed with error: {}",
-                        self.dst_client_id, self.dst_chain_id, e
-                    ))
-                    .exit()
-                }
-            };
+        let src_chain_id = dst_chain
+            .query_client_state(&self.dst_client_id, ibc::Height::zero())
+            .map_err(|e| {
+                format!(
+                    "Query of client '{}' on chain '{}' failed with error: {}",
+                    self.dst_client_id, self.dst_chain_id, e
+                )
+            })?
+            .chain_id();
 
-        let src_chain = match spawn_chain_runtime(&config, &src_chain_id) {
-            Ok(handle) => handle,
-            Err(e) => return Output::error(format!("{}", e)).exit(),
-        };
+        let src_chain =
+            spawn_chain_runtime(&config, &src_chain_id).map_err(|e| format!("{}", e))?;
 
-        let height = match self.target_height {
-            Some(height) => ibc::Height::new(src_chain.id().version(), height),
-            None => ibc::Height::zero(),
-        };
+        let src_chain_id = src_chain.id().version();
+
+        let client = ForeignClient::find(src_chain, dst_chain, &self.dst_client_id)
+            .map_err(|e| format!("failed to find foreign client: {}", e))?;
 
-        let trusted_height = match self.trusted_height {
-            Some(height) => ibc::Height::new(src_chain.id().version(), height),
-            None => ibc::Height::zero(),
+        let m_target_height = self
+            .target_height
+            .and_then(|h| NonZeroHeight::new(ibc::Height::new(src_chain_id, h)));
+
+        let m_trusted_height = self
+            .trusted_height
+            .and_then(|h| NonZeroHeight::new(ibc::Height::new(src_chain_id, h)));
+
+        let target_height = match m_target_height {
```

Comment:
These checks make the command quite intelligent, which is nice. But it might be better if the user is allowed to exercise the corner-cases of `update client CLIENT-ID --target-height 0` for example, especially for `tx raw`. To do that, it could be a good idea to leave `TxUpdateClientCmd` untouched (modulo the nice refactor with `execute`) and create a new command e.g., `UpdateClientCmd` that is similar to `TxUpdateClientCmd ` but which has more checks in place to cover corner-cases.

Having a new `UpdateClientCmd` command will also address (part of) #773 .

Reply:
I have added a `NonZeroHeight::unsafe_new()` function for creation of heights that are really meant to be zero. Btw note that previously even if zero height is allowed in the command line, it still really means the absence of height and the relayer will still fetch the current height.

I have now added the `UpdateClientCmd` and have the two behaviors slightly different. In `UpdateClientCmd`, only the target height can optionally be provided, and the command will wait for the target height to reach before updating. For `TxUpdateClientCmd`, both trusted height and target height must be provided, and there is no sanity checking in the command. So if invalid heights are provided, they will go into the implementation code and return error from there.

Reply:
This is just my suggestion, because otherwise I am not really sure what is the point of having two commands with separate implementations if their behavior are the same. Or do you have any recommendation of how the two commands should be differentiated?